lib/msal-core/src/UserAgentApplication.ts

```diff
@@ -1004,105 +1011,103 @@ export class UserAgentApplication {
   acquireTokenSilent(scopes: Array<string>, authority?: string, user?: User, extraQueryParameters?: string): Promise<string> {
     return new Promise<string>((resolve, reject) => {
       // Validate scopes
-      // TODO: Replace with new validation pattern
-      const isValidScope = this.validateInputScope(scopes);
-      if (isValidScope && !Utils.isEmpty(isValidScope)) {
+      try {
+        scopes = this.validateAndFilterInputScope(scopes);
+      } catch (err) {
+        if (err instanceof ClientConfigurationError) {
+          // These will return errors instead of strings in the future
+          // Should we throw instead of reject here?
+          reject(err.message);
+        } else {
+          reject(ErrorCodes.inputScopesError + Constants.resourceDelimeter + ErrorDescription.inputScopesError);
+        }
+      }
+
+      const scope = scopes.join(" ").toLowerCase();
+      const userObject = user ? user : this.getUser();
+      const adalIdToken = this._cacheStorage.getItem(Constants.adalIdToken);
+      //if user is not currently logged in and no login_hint/sid is passed as an extraQueryParamater
+      if (!userObject && Utils.checkSSO(extraQueryParameters) && Utils.isEmpty(adalIdToken) ) {
+        this._logger.info("User login is required");
         // TODO: Reject with custom error here
-        // TODO: Is this always accessToken?
-        reject(ErrorCodes.inputScopesError + "|" + isValidScope);
+        reject(ErrorCodes.userLoginError + Constants.resourceDelimeter + ErrorDescription.userLoginError);
         return null;
-      } else {
-        // TODO: Remove this from the else block, it is unnecessary
-        if (scopes) {
-          scopes = this.filterScopes(scopes);
-        }
+      }
+      //if user didn't passes the login_hint and adal's idtoken is present and no userobject, use the login_hint from adal's idToken
+      else if (!userObject && !Utils.isEmpty(adalIdToken)) {
+        const idTokenObject = Utils.extractIdToken(adalIdToken);
+        console.log("ADAL's idToken exists. Extracting login information from ADAL's idToken ");
+        extraQueryParameters = Utils.constructUnifiedCacheExtraQueryParameter(idTokenObject, extraQueryParameters);
+      }
 
-        const scope = scopes.join(" ").toLowerCase();
-        const userObject = user ? user : this.getUser();
-        const adalIdToken = this._cacheStorage.getItem(Constants.adalIdToken);
-        //if user is not currently logged in and no login_hint/sid is passed as an extraQueryParamater
-        if (!userObject && Utils.checkSSO(extraQueryParameters) && Utils.isEmpty(adalIdToken) ) {
-          this._logger.info("User login is required");
-          // TODO: Reject with custom error here
-          reject(ErrorCodes.userLoginError + Constants.resourceDelimeter + ErrorDescription.userLoginError);
-          return null;
+      let authenticationRequest: AuthenticationRequestParameters;
```

Comment:
lets can sometimes be good examples of things to factor out.

instead of let and then follow it getting filled, let the logic look like this
```js
const authenticationRequest: AuthenticationRequestParameters = constructAuthRequest(whateverIneed);
```

this then becomes really easy to test right, we have a function that we expect given certain inputs to produce certain results instead of in a big file, hoping the right thing is happening

Reply:
Please create an item in VSTS for this. I am resolving this here.